### ISSUE TITLE
Suggestion for a corrupted file fix

### ISSUE THREAD

**Issue by HoloWise** [yellowbluesky/PixivforMuzei3#76]
Hey, I've been thinking about the issue I mentioned last time about some of the files being corrupted and since it kinda annoys me (yeah ik I said it's not a big deal lol) I thought I may look around and suggest a fix. :P I'm not sure personally how reliable it is but you could check if downloaded files follow a proper structure of the format they were originally in. I never had to deal with it in my apps so I really have no practical experience with image formats.

For example a JPEG file should start with 0xFF, 0xD8 and end with 0xFF, 0xD9 

(I used artwork 79521139 for that example which is stored by pixiv as JPG)

example of start:
![obraz](https://user-images.githubusercontent.com/36460595/74793357-629ee380-52c0-11ea-9440-e9dccb377108.png)
example of end:
![obraz](https://user-images.githubusercontent.com/36460595/74793378-721e2c80-52c0-11ea-9742-138128712d25.png)

For PNG you can look for IHDR (0x49 0x48 0x44 0x52) which should be the first header of a png file and IEND (0x49 0x45 0x4e 0x44) which marks the end of PNG datastream.

(I used artwork 79559763 for that example which is stored by pixiv as PNG)

example of start:
![obraz](https://user-images.githubusercontent.com/36460595/74793633-2d46c580-52c1-11ea-83cc-88d49495989b.png)
example of end:
![obraz](https://user-images.githubusercontent.com/36460595/74793652-40599580-52c1-11ea-8597-043e18dbdadb.png)

Obviously things aren't as simple as that and for example you can encounter SOI SOI EOI EOI in jpg because it has embedded thumbnail thus making it possible to find EOI (0xFF, 0xD9) prematurely if you aren't checking length specifiers, skipping segments etc. Obviously at this point it starts to get unnecessary complicated and would require a properly written file validator that can do all this stuff (and there are other variables such as how pixiv modifies files etc.) but I thought that maybe you could add a simple (and flawed) version that checks for end of file for a specific file format and attempts a redownload (mb with delay if it's connection issue?) if it couldn't find it during the entire process. It would catch the image I used as an example when I first brought it up 2 weeks ago since it didn't contain EOI. 

Also I'm not sure how that kind of verification would work for gif files and it's getting a bit late so I may have to check it tomorrow.

**Comment by yellowbluesky** [comment]
Thanks for your investigation :)
I can add this in, it looks straightforward enough to just check the last few bytes for the presence of a magic byte sequence
I'll add in functionality for GIF, I can research it, but I've never seen a GIF in Pixiv

**Comment by yellowbluesky** [comment]
I had a quick look, none of the JPEG's I looked at had embedded thumbnails, so I can naively search for any FF D9 byte sequence

**Comment by HoloWise** [comment]
Sounds like it may work then :P 

I mentioned gif cuz ik your code recognizes jpg, png and gif when downloading images (altho idk how that works with their weird animations that are just a series of pngs as far as I remember)

I did look it up tho

1. jpg ends with FF D9

2. png ends with 49 45 4E 44 (IEND) and then 4-byte CRC (Cyclic Redundancy Check) - "calculated on the preceding bytes in the chunk, including the chunk type code and chunk data fields, but not including the length field. The CRC is always present, even for chunks containing no data"

So since IEND has no data I assume CRC will always be the same (and it was in the images I checked)

3. In gifs 3B at the end is a file terminator

**Comment by HoloWise** [comment]
Did you change file format that your app uses to save files? I remember you said you change everything to png. I realised I have a bunch of jpg files in my folders downloaded as recently as today. I'm on version 2.7.1 


Btw. Kinda on a side note cuz it's a different use case but I wanted to check how reliable checking last bytes is so I wrote an app that passes all 1627 images i got through the app so far through a method that checks last 10 bytes and sorts based on what sequence of bytes it finds at the end and separates corrupted files. One corrupted jpg ended up being marked as gif cuz it ended with 3B but that wouldn't be the case for you since you can pick a way you check last bytes based on original extension instead of passing a file through all 3. Disregarding that file, all files were properly sorted to their file types regardless of edited extension (so all jpgs with png extension were moved to JPG folder etc. verified with TrID Identifier) and none of the corrupted files ended up getting through (verified with Bad Peggy) and were separated to a corrupted folder. So it seems like checking last bytes is a pretty reliable way of verifying if a file is corrupted. (I used FF D9 for JPG check and 49 45 4E 44 AE 42 60 82 (4 bytes IEND and 4 bytes CRC) for PNG check but I can't promise that it'll always work cuz idk if CRC for IEND is always the same (tho logically it should be lol)

**Comment by yellowbluesky** [comment]
Hi there,

I did change the file extension to .jpg. The actual file type of the image is unchanged, and this was an arbitrary choice really, not sure why I made it.

The images come from Pixiv as a raw bytestream, and there was previously no easy ways to check file type (or I wasn't bothered to), so they all got stuck with a default type that I decided on. I don't convert image types.

With the additions I'm adding now to check for corrupt images, I can also properly assign the correct file type for each image.

I've got a prototype for corrupt image checking working (not sure if it's up on the repo), I can likely have something working up tonight (in hours or so)

Thanks for the research you've done on file structures for each of the types, it's handy to have a reference all in one spot

**Comment by HoloWise** [comment]
Yeah I know you don't convert them ^^ My bad for calling it a file format and not extension.

> With the additions I'm adding now to check for corrupt images, I can also properly assign the correct file type for each image.

Idk the code that well but I assumed use of IMAGE_SUFFIXS in getRemoteFileExtension implies you already know the extensions of files you download. Guess I was wrong to assume that :P I should look a bit more into the code at some point.

I'm looking forward to your implementation of file verification.

**Comment by yellowbluesky** [comment]
https://github.com/yellowbluesky/PixivforMuzei3/releases/tag/2.7.2

Try this release.

I've not tested how the app behaves when it detects a corrupt image. It should retry to download a new picture after a short delay (10 seconds I think), but I've not been able to trigger a corrupt image manually to test it myself.

let me know if you have any abnormal behavior or problems in my app

**Comment by yellowbluesky** [comment]
BTW, I actually forgot about getRemoteFileExtensions(), it's been a long week for me haha

That function is used only for downloading images from one of the rankings:

- Ranking JSON's don't actually give me the URL to the full resolution image, only a low res thumbnail which is always a JPG
- I need to determine the correct URL for the high resolution image, and I do that by brute force pinging each combination of file extension and URL until I get a response that isn't a 404 error

Now that you mention it, I could have probably used this information to aid ranking file extension assignment.

Another thing I just realised thanks to you, all the auth feed modes provide me with the correct file extensions already in their JSON's, meaning I didn't need to kludge together my curent method of checking bytes haha

I'll likely make changes to how I detect corrupt fiels and file extensions, I'm not super happy about it; it's inefficient and a bit ugle

**Comment by HoloWise** [comment]
Seems like it's good that I mentioned it then lol 

Just installed 2.7.2 so I'll let you know if I see any abnormalities with how the app handles corrupted files in it's entirety but since I had those 1627 files from yesterday on my pc I copied isImageCorrupt method and passed them through it (I added a file type counter and replaced reading to byte array through FileUtils from Commons IO with simple Files.readAllBytes(Path path) cuz I'm too lazy to add Commons IO lol). Got exact same result as the app I wrote yesterday when it comes to number of jpgs, pngs and corrupted files. Ended up adding some code to compare arrays of names sorted by isImageCorrupt with sorting I did by reading last 10 bytes and arrays were the same as well so isImageCorrupt does behave as expected (Those 1627 images date back to when I first asked you to implement saving to external storage)

**Comment by yellowbluesky** [comment]
Nice to see that it works just like you expected it to

Just for reference, can I see how you coded your file integrity check?

**Comment by HoloWise** [comment]
Code: https://pastebin.com/cBEJm5Mn

There was a big comment discussing this approach so I removed it to not make that paste huge but if you want to read through it then here https://pastebin.com/JhGUvLuy

I wanted to share it through github but code tag was messing it up for some reason lol.

**Comment by yellowbluesky** [comment]
Thanks for your code, and thanks for the giant write up haha

I'll close the issue now, feel free to reopen it if you encounter a problem

**Comment by HoloWise** [comment]
Ok so I encountered one issue but I'm not sure if it's connected to introduced changes. I have the app set to 1am cache clear and I checked the gallery through muzei at ~1:30am. It had 23 of those corrupted images that I talked about in the past. (not corrupted files, all files in pixiv folder are fine so far). No thumbnail, on press shows name of an artwork but doesn't change current wallpaper. But unlike in the past it didn't completely brick muzei, after those 23 images it started loading images properly and continues to work properly.
